workshops/views.py

```diff
@@ -145,51 +148,201 @@ def person_bulk_add(request):
+            if any_errors:
+                messages.add_message(request, messages.ERROR,
+                                     "Please make sure to fix all errors "
+                                     "listed below.")
+
+            context = {'title': 'Confirm uploaded data',
+                       'persons_tasks': persons_tasks}
+            return render(request, 'workshops/person_bulk_add_results.html',
+                          context)
+
+        elif (request.POST.get('confirm', None) and
+                not request.POST.get('cancel', None)):
+            # there must be "confirm" and no "cancel" in POST in order to save
+
+            try:
+                records = 0
+                with transaction.atomic():
+                    for row in persons_tasks:
+                        # create person
+                        p = Person(**row['person'])
+                        p.save()
+                        records += 1
+
+                        # create task if data supplied
+                        if row['event'] and row['role']:
+                            e = Event.objects.get(slug=row['event'])
+                            r = Role.objects.get(name=row['role'])
+                            t = Task(person=p, event=e, role=r)
+                            t.save()
+                            records += 1
+
+            except (IntegrityError, ObjectDoesNotExist) as e:
+                messages.add_message(request, messages.ERROR,
+                                     "Error saving data to the database: {}. "
+                                     "Please make sure to fix all errors "
+                                     "listed below.".format(e))
+                _verify_upload_person_task(persons_tasks)
+                context = {'title': 'Confirm uploaded data',
+                           'persons_tasks': persons_tasks}
+                return render(request,
+                              'workshops/person_bulk_add_results.html',
+                              context)
+
+            else:
+                request.session['bulk-add-people'] = None
+                messages.add_message(request, messages.SUCCESS,
+                                     "Successfully bulk-loaded {} records."
+                                     .format(records))
+                return redirect('person_bulk_add')
+
+        else:
+            # any "cancel" or no "confirm" in POST cancels the upload
+            request.session['bulk-add-people'] = None
+            return redirect('person_bulk_add')
+
+    else:
+        # alters persons_tasks via reference
+        _verify_upload_person_task(persons_tasks)
+
+        context = {'title': 'Confirm uploaded data',
+                   'persons_tasks': persons_tasks}
+        return render(request, 'workshops/person_bulk_add_results.html',
+                      context)
+
+
 def _upload_person_task_csv(request, uploaded_file):
```

Comment:
Mostly a style preference but I generally prefer to keep view functions (those returning HttpResponses) in `views.py` and non-view functions (those that don't) into a `<app_name>.utils.py` file to keep `views.py` from becoming a huge mess. Such a change would apply to  `_verify_upload_person_task` as well.

Then, when in testing, we can directly import the util functions and test them and then test the views with `TestCase.Client`.